File: Data Structures & Algorithms/contains-duplicate-ii/submission-1.java
class Solution {
    public boolean containsNearbyDuplicate(int[] arr, int k) {
        int n = arr.length;

        int left = 0;

        for (int right = 1 ; right < n ; right++) {
            if (right - left <= k && arr[right] == arr[left]) {
                return true;

            } else if (right - left <= k) {
                right++;

            } else {
                left = right;
            }
        }

        return false;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        n = len(piles)
        left, right = min(piles), max(piles)

        def isValid(piles, mid, h):
            totalhours = 0
            for pile in piles:
                totalhours += math.ceil(pile / mid)
                if totalhours > h:
                    return False
            return True
        
        ans = 0
        while left <= right:
            mid = left + (right - left) // 2
            if isValid(piles, mid, h):
                ans = mid
                right = mid - 1
            else:
                left = mid + 1

        return ans